src/t/d4.clj

```diff
@@ -0,0 +1,164 @@
+(ns t.d4
+  (:gen-class)
+  (:require
+    [clojure.string :as cljstr]
+    [clojure.set :as cljset]))
+
+(def sample
+"MMMSXXMASM
+MSAMXMSMSA
+AMXSXMAAMM
+MSAMASMSMX
+XMASAMXAMM
+XXAMMXXAMA
+SMSMSASXSS
+SAXAMASAAA
+MAMMMXMMMM
+MXMXAXMASX")
+
+(defn parse-input
+  [input]
+  (->> input
+       cljstr/split-lines
+       (map-indexed (fn [i l]
+                      (->> l
+                      (map-indexed (fn [ii letter] {letter [[i ii]]}))
+                      (apply merge-with into))))
+       (apply merge-with into)))
+
+
+(comment
+  all this is useless because I misunderstood the thing that explains the stuff
+(defn search-letter-around
+  [pos letter-pos]
+  (let [x (first pos)
+        y (second pos)
+        around (for [i [-1 0 +1]
+                     ii [-1 0 +1]]
+                 [(+ x i) (+ y ii)])]
+    (cljset/intersection letter-pos (set around))))
+
+(defn loop-on-S
+  [Apos Ss]
+  (count (search-letter-around Apos Ss)))
+
+(defn loop-on-A
+  [Mpos As Ss]
+  (let [as (search-letter-around Mpos As)]
+   (loop [a (first as)
+          otherAs (rest as)
+          tot 0]
+     (if (nil? a)
+       tot
+       (recur (first otherAs) (rest otherAs) (+ tot
+                                                (loop-on-S a Ss)))))))
+
+(defn loop-on-M
+  [Xpos Ms As Ss]
+  (let [ms (search-letter-around Xpos Ms)]
+   (loop [m (first ms)
+          otherMs (rest ms)
+          tot 0]
+     (if (nil? m)
+       tot
+       (recur (first otherMs) (rest otherMs) (+ tot
+                                                (loop-on-A m As Ss)))))))
+(defn d4p1
+  [input]
+  (let [p-input (parse-input input)
+        Xs (set (get p-input \X))
+        Ms (set (get p-input \M))
+        As (set (get p-input \A))
+        Ss (set (get p-input \S))
+        ]
+    (loop [x (first Xs)
+           otherXs (rest Xs)
+           tot 0]
+      (if (nil? x)
+        tot
+      (recur (first otherXs) (rest otherXs) (+ tot
+                                                (loop-on-M x Ms As Ss)))))))
+)
+
+(defn xmas?
+  [xpos dir Ms As Ss]
+  (let [x (first xpos)
+        y (second xpos)
+        dx (first dir)
+        dy (second dir)]
+    (and
+      (contains? Ms [(+ x dx)       (+ y dy)])
+      (contains? As [(+ x dx dx)    (+ y dy dy)])
+      (contains? Ss [(+ x dx dx dx) (+ y dy dy dy)])
+      )))
+
+(defn d4p1
+  [input]
+  (let [p-input (parse-input input)
+        Xs (set (get p-input \X))
+        Ms (set (get p-input \M))
+        As (set (get p-input \A))
+        Ss (set (get p-input \S))
+        dirs (for [dx [+1 0 -1] dy [+1 0 -1]] [dx dy])
+        ]
+    (loop [x (first Xs)
+           otherXs (rest Xs)
+           tot 0]
+      (if (nil? x)
+        tot
+        (let [tots (map (fn [dir] (if (xmas? x dir Ms As Ss) 1 0)) dirs)]
+          (recur (first otherXs) (rest otherXs) (apply + tot tots)))
+      ))))
```

Comment:
This looks like it could be a reduce:
```suggestion
    (reduce (fn [tot x] (+ tot (count (keep (fn [dir] (xmas? x dir Ms As Ss)) dirs))))
                  0
                  Xs)))
```

Reply:
oh yeah, that is probably right. I thought in term of loop and expected complexe operation within ...